### PR TITLE
Support 4bits quantization

### PR DESCRIPTION
**Pull request by hans00**
As title.

Notes:
- Execution require `onnxruntime>=1.16.2`
- Convert require `onnxruntime>=1.17.0`
- 4bits currently support CPU / CUDA / WebGPU(future) ([MatMulNBits](https://github.com/search?q=repo%3Amicrosoft%2Fonnxruntime+MatMulNBits&type=code))
- bnb4 currently support CPU / CUDA ([MatMulBnb4](https://github.com/search?q=repo%3Amicrosoft%2Fonnxruntime+BNB4&type=code))

### CODE REVIEW

**Comment by xenova** [comment]
Thanks so much! This is amazing! I think it would be good to also name/suffix these models slightly differently, so you can export (and use) multiple versions.

**Comment by bil-ash** [comment]
@hans00 Can the bnb4 quantization be used with webgpu backend?

**Comment by xenova** [comment]
@bil-ash ORT recently added support for `MatMulNBit` and will be available in v1.17.3. You can already access it with [this version](https://www.npmjs.com/package/onnxruntime-web/v/1.18.0-dev.20240309-efad5bbc5a).

File: scripts/convert.py [hunk 1]
```diff
@@ -158,6 +165,12 @@ class ConversionArguments:
             "help": "Whether to quantize the model."
         }
     )
+    quantize_mode: str = field(
+        default=QuantMode.BIT8,
+        metadata={
+            "help": "Quantization mode to use. Options are: int4, int8, bnb4"
```
Comment (xenova) [inline scripts/convert.py:171]: mismatched names between this and `QuantMode.*`

**Comment by hans00** [comment]
> Also, to help test, would you like to create a demo model on the HF Hub which contains all these different variants?

Here is my convertion:
https://huggingface.co/BricksDisplay/stablelm-2-1_6b-q4
https://huggingface.co/BricksDisplay/stablelm-2-1_6b-bnb4
https://huggingface.co/BricksDisplay/phi-1_5-q4
https://huggingface.co/BricksDisplay/phi-1_5-bnb4

As my test (Windows with DML, CPU EP) using StableLM-2-1.6b
Q8: 0.95 token/s
Q4: 2.46 token/s
BNB4: 0.44 token/s

**Comment by hans00** [comment]
For benchmark, the quantized `all-MiniLM-L6-v2` model is
https://huggingface.co/BricksDisplay/all-MiniLM-L6-v2-q4
https://huggingface.co/BricksDisplay/all-MiniLM-L6-v2-bnb4

**Comment by xenova** [comment]
> Did you mean the model file name like model_quantized_int4.onnx or model_quantized_nf4.onnx ?

Yes, exactly :) I think this PR can help standardize which suffixes we should use (as `quantized` will be overloaded). It might be worthwhile aligning this with other libraries' naming conversions. Do you have any thoughts on this?

Also, just checking, but the [bnb4 version](https://huggingface.co/BricksDisplay/all-MiniLM-L6-v2-bnb4/tree/main/onnx) has model sizes which are the same as fp16 (only half the size of the original). Is this correct?

**Comment by hans00** [comment]
> Yes, exactly :) I think this PR can help standardize which suffixes we should use (as quantized will be overloaded). It might be worthwhile aligning this with other libraries' naming conversions. Do you have any thoughts on this?

I have no idea.
My naming will be `{orig_name}_quantized_{quant_type}.onnx`

> Also, just checking, but the [bnb4](https://huggingface.co/BricksDisplay/all-MiniLM-L6-v2-bnb4/tree/main/onnx) and [q4](https://huggingface.co/BricksDisplay/all-MiniLM-L6-v2-q4/tree/main/onnx) versions have model sizes which are the same as fp16 (only half the size of the original). Is this correct?

The dtype is not change, just convert from `sentence-transformers/all-MiniLM-L6-v2`.
The `MatMul` op should quantize to INT4 or NF4, others unchanged (maybe FP32)

File: scripts/convert.py [hunk 2]
```diff
@@ -5,6 +5,8 @@
 from dataclasses import dataclass, field
 from typing import Optional, Set
 from tqdm import tqdm
+from enum import Enum
+from pathlib import Path
```
Comment (xenova) [inline scripts/convert.py:9]:
```suggestion
```

File: scripts/convert.py (hunk 1 above)
Comment (xenova) [reply scripts/convert.py:174]: The names are still mismatched 👀 (int4 and int8, specifically)

File: scripts/convert.py [hunk 3]
```diff
@@ -270,11 +286,13 @@ def quantize(model_names_or_paths, **quantize_kwargs):
 
     quantize_config = dict(
         **quantize_kwargs,
+        quantize_mode=mode.value,
```
Comment (xenova) [inline scripts/convert.py:289]: This would prevent multiple quantizations to be produced per model.

**Comment by xenova** [comment]
Thanks for this contribution 🤗! Since 4-bit quantization will only be supported in v3, I've merged it into [that branch](https://github.com/xenova/transformers.js/pull/545/commits/9efbada2b97fdb22fb0775ebbe30d93371cb83a3).

**Comment by xenova** [comment]
@hans00 Have you by any chance experimented with q4f16 formats (model is and runs in fp16, with int4 block weight quantization)? That's the difference between:
- (fp32) https://huggingface.co/Xenova/Phi-3-mini-4k-instruct, and
- (fp16) https://huggingface.co/Xenova/Phi-3-mini-4k-instruct_fp16

**Pull request merged**